dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageChecker.java

```diff
@@ -303,11 +303,46 @@ JarLinkageReport generateLinkageReport(Path jarPath, SymbolReferenceSet symbolRe
       ClassSymbolReference reference) {
     String targetClassName = reference.getTargetClassName();
     try {
-      classDumper.loadJavaClass(targetClassName);
+      JavaClass targetClass = classDumper.loadJavaClass(targetClassName);
+      if (!isClassAccessibleFrom(targetClass, reference.getSourceClassName())) {
+        return Optional.of(LinkageErrorMissingClass.errorInvalidModifier(
+            classDumper.findClassLocation(targetClassName), reference));
+      }
       return Optional.empty();
     } catch (ClassNotFoundException ex) {
-      return Optional.of(LinkageErrorMissingClass.errorAt(reference));
+      return Optional.of(LinkageErrorMissingClass.errorMissingTargetClass(reference));
+    }
+  }
+
+  /**
+   * Returns true if the {@code javaClass} is accessible {@code from sourceClassName} in terms of
+   * the access modifiers in the {@code javaClass}.
+   *
+   * @see <a href="https://docs.oracle.com/javase/specs/jls/se8/html/jls-8.html#jls-ClassModifier">
+   *   JLS 8.1.1. Class Modifiers</a>
+   */
+  private static boolean isClassAccessibleFrom(JavaClass javaClass, String sourceClassName) {
+    if (javaClass.isPublic()) {
+      // If a class is public, then it is accessible from anywhere
```

Comment:
This analysis is wrong in the case of nested classes.

Reply:
Fixed with recursively calling this for each enclosing class.